aspnet/fundamentals/hosting.rst

```diff
@@ -0,0 +1,233 @@
+
+.. code-block:: c#
+
+    new WebHostBuilder()
+        .UseStartup("StartupAssemblyName")
+
+.. _web-root-setting:
+
+Web Root ``string``
+    Key: ``webroot``. If not specified the default is ``(Content Root Path)\wwwroot``, if it exists. If this path doesn't exist, then a no-op file provider is used. Set using ``UseWebRoot``.
+
+.. code-block:: c#
+
+    new WebHostBuilder()
+        .UseWebRoot("public")
+
+Use :doc:`configuration` to set configuration values to be used by the host. These values may be subsequently overridden. This is specified using ``UseConfiguration``.
+
+.. code-block:: c#
+  :emphasize-lines: 3-6,9
+
+  public static void Main(string[] args)
+  {
+    var config = new ConfigurationBuilder()
+      .AddCommandLine(args)
+      .AddJsonFile("hosting.json", optional: true)
+      .Build();
+
+    var host = new WebHostBuilder()
+      .UseConfiguration(config)
+      .UseKestrel()
+      .Configure(app => 
+      {
+        app.Run(async (context) => await context.Response.WriteAsync("Hi!"));
+      })
+    .Build();
+
+    host.Run();
+  }
+
+In the example above, command line arguments may be passed in to configure the host, or configuration settings may optionally be specified in a *hosting.json* file. To specify the host run on a particular URL, you could pass in the desired value from the command line:
+
+.. code-block:: none
+
+  dotnet run --urls "http://*:5000"
+
+The ``Run`` method starts the web app and blocks the calling thread until the host is shutdown.
+
+.. code-block:: c#
+
+    host.Run();
+
+You can run the host in a non-blocking manner by calling its ``Start`` method:
+
+.. code-block:: c#
+
+  using (host)
+  {
+    host.Start();
+    Console.ReadLine();
+  }
+
+Pass a list of URLs to the ``Start`` method and it will listen on the URLs specified:
+
+.. code-block:: c#
+
+  var urls = new List<string>() { 
+    "http://*:5000", 
+    "http://localhost:5001"
+    };
+  var host = new WebHostBuilder()
+    .UseKestrel()
+    .UseStartup<Startup>()
+    .Start(urls.ToArray());
+  
+  using (host)
+  {
+    Console.ReadLine();
+  }
+
+Ordering Importance
+^^^^^^^^^^^^^^^^^^^
+
+``WebHostBuilder`` settings are first read from certain environment variables, if set. These environment variables must use the format ``ASPNETCORE_{configurationKey}``, so for example to set the URLs the server will listen on by default, you would set ``ASPNETCORE_URLS``.
+
+You can override any of these environment variable values by specifying configuration (using ``UseConfiguration``) or by setting the value explicitly (using ``UseUrls`` for instance). The host will use whichever option sets the value last, so for instance if you wanted to be programmatically set the default URL to one value, but allow it to be overridden with configuration, you could configure the host as follows:
```

Comment:
RE: For this reason UseIISIntegration must come after UseUrls because it replaces the url with one dynamically provided by IIS/ANCM.